src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoader.java

```diff
@@ -32,43 +37,129 @@
+                    // Otherwise only add text segments outside the invocation as the invocation should only contain the end variable
+                    else if (nestedInvocationCount == 0) {
+                        simplifiedTemplateSegments.add(templateSegment);
+                    }
+                }
+                // If we find a variable not inside an invocation, we can't simplify this template
+                else if ((templateSegment.isVariable() || templateSegment.isNamedVariable()) && (nestedInvocationCount == 0)) {
+                    hasVariablesOutsideInvocation = true;
+                    break;
+                }
+            }
 
+            // If we found an invocation (including balanced parens) and all variables were inside of it, use the simplified template segments
+            if (hasInvocation && (nestedInvocationCount == 0) && !hasVariablesOutsideInvocation) {
+                effectiveTemplateSegments = simplifiedTemplateSegments;
+            }
+        }
+
+        // Build the template
+        for (SnippetTemplateSegment templateSegment : effectiveTemplateSegments) {
+            if (templateSegment.isTextSegment()) {
+                template.addTextSegment(templateSegment.getTextSegment());
+            } else if (templateSegment.isVariableSegment()) {
+                template.addVariableSegment(templateSegment.getVariableSegment());
+            } else if (templateSegment.isVariable()) {
+                template.addVariable(templateSegment.getVariable(), templateSegment.isAlwaysStopAt());
+            } else if (templateSegment.isNamedVariable()) {
+                template.addVariable(
+                        templateSegment.getVariableName(),
+                        templateSegment.getVariable(),
+                        templateSegment.getDefaultValueExpression(),
+                        templateSegment.isAlwaysStopAt(),
+                        templateSegment.isSkipOnStart()
+                );
+            } else if (templateSegment.isEndVariable()) {
+                template.addEndVariable();
+            }
+        }
+    }
+
+    // NOTE: These are quite simple now and basically look for a segment that ends with "(" (ignoring trailing white
+    // space) to determine that an invocation has started and one that begins with ")" (again, ignore leading white
+    // space) to determine that an invocation has ended. If LSP languages with more complex invocation patterns are
+    // found these will need to be updated accordingly.
+
+    private static boolean isInvocationStart(@NotNull String text) {
+        return StringUtil.trimTrailing(text).endsWith("(");
+    }
+
+    private static boolean isInvocationEnd(@NotNull String text) {
+        return StringUtil.trimLeading(text).startsWith(")");
     }
 
     @Override
     public void text(String text) {
-        template.addTextSegment(formatText(text));
+        templateSegments.add(SnippetTemplateSegment.textSegment(formatText(text)));

```

Comment:
From here down, what you'll see is the change from building the template itself to accumulating what's needed to build the template in `endSnippet()`.